### PR TITLE
Add support for pivot table and entity in manyToMany relationships

### PR DESCRIPTION
**Pull request by thetutlage**
FIXES CORE-2349

This PR adds support for defining custom `pivotTable` and `pivotEntity` for a `manyToMany` relationship

### CODE REVIEW

**Review by olivermrbl — APPROVED** [review]
LGTM, nice work

**Review by riqwan — APPROVED** [review]
🔥

**Comment by thetutlage** [comment]
I will wait for @adrien2p to approve it as well :)

File: packages/core/utils/src/dml/__tests__/entity-builder.spec.ts [hunk 1]
```diff
@@ -3023,7 +3023,7 @@ describe("Entity builder", () => {
       const user = model.define("user", {
         id: model.number(),
         username: model.text(),
-        email: model.manyToMany(() => email),
+        email: model.belongsTo(() => email),
```
Comment (adrien2p) [inline packages/core/utils/src/dml/__tests__/entity-builder.spec.ts:3026]: nice catch 🙏

File: packages/core/utils/src/dml/entity-builder.ts [hunk 2]
```diff
@@ -162,7 +162,13 @@ export class EntityBuilder {
    *   relationship requires a pivot table to establish a many to many
    *   relationship between two entities
    */
-  manyToMany<T>(entityBuilder: T, options?: RelationshipOptions) {
+  manyToMany<T>(
+    entityBuilder: T,
+    options?: RelationshipOptions & {
+      pivotTable?: string
+      pivotEntity?: () => DmlEntity<any>
+    }
```
Comment (adrien2p) [inline packages/core/utils/src/dml/entity-builder.ts]:
```suggestion
    options?: RelationshipOptions & ({
      pivotTable?: string
      pivotEntity?: never
    } | {
      pivotTable?: never
      pivotEntity?: () => DmlEntity<any>
    })
```

The idea being that we only allow one or the other wdyt? Unless it makes sense to have both but not sure

**Review by adrien2p — APPROVED** [review]
LGTM, few suggestions

File: packages/core/utils/src/dml/entity-builder.ts (hunk 2 above)
Comment (thetutlage) [reply packages/core/utils/src/dml/entity-builder.ts]: What if I want to use a custom pivot table name with a custom pivot entity. Remember, we do not allow defining a custom table name via entities. So there is no way for me to configure that on the pivot entity

Comment (adrien2p) [reply packages/core/utils/src/dml/entity-builder.ts]: yes thats true, but don't we want to have the same behaviour has any other entity? meaning that the table name will be constructed from the entity name, here the pivot entity name would be the one defining this table name. If you look at the migration level it is also probably what will happen, a table will be created for the pivot entity which might in that case defer from configuration

Comment (adrien2p) [reply packages/core/utils/src/dml/entity-builder.ts]: Validated through a call discussion 💪

Comment (thetutlage) [reply packages/core/utils/src/dml/entity-builder.ts]: Makes sense!

**Pull request merged**